### ISSUE TITLE
multiple applications in tomcat and jruby 9.4.3.0 com.headius.invokebinder.InvalidTransformException: java.lang.IllegalAccessException

### ISSUE THREAD

**Issue by jlahtinen** [jruby/jruby#7911]
Starting multiple rails applications with jruby 9.4.3.0 eventually fails with error

    java.lang.NoClassDefFoundError: Could not initialize class org.jruby.ir.targets.indy.Bootstrap
            at org.jruby.ir.targets.indy.ConstantLookupSite.searchConst(ConstantLookupSite.java:107)

After moving all static field initialization (that is the reason line numbers in Bootstrap.java in next strace do not match with tag 9.4.3.0) to one static block and rescuing exception there and loggin it:

    com.headius.invokebinder.InvalidTransformException: java.lang.IllegalAccessException: no such field: org.jruby.runtime.ThreadContext.runtime/org.jruby.Ruby/getField
      at com.headius.invokebinder.Binder.getFieldQuiet(Binder.java:1757)
      at org.jruby.ir.targets.indy.Bootstrap.<clinit>(Bootstrap.java:215)
      at org.jruby.ir.targets.indy.IndyValueCompiler.pushNil(IndyValueCompiler.java:133)
      at org.jruby.ir.targets.simple.NormalLocalVariableCompiler.getHeapLocalOrNil(NormalLocalVariableCompiler.java:42)
      at org.jruby.ir.targets.JVMVisitor.LocalVariable(JVMVisitor.java:2751)
      at org.jruby.ir.targets.JVMVisitor.ClosureLocalVariable(JVMVisitor.java:2656)
      at org.jruby.ir.operands.ClosureLocalVariable.visit(ClosureLocalVariable.java:49)
      at org.jruby.ir.targets.JVMVisitor.visit(JVMVisitor.java:636)
      at org.jruby.ir.targets.JVMVisitor.compileCallCommon(JVMVisitor.java:1235)
      at org.jruby.ir.targets.JVMVisitor.CallInstr(JVMVisitor.java:1225)
      at org.jruby.ir.instructions.CallInstr.visit(CallInstr.java:150)
      at org.jruby.ir.targets.JVMVisitor.visit(JVMVisitor.java:632)
      at org.jruby.ir.targets.JVMVisitor.emitScope(JVMVisitor.java:305)
      at org.jruby.ir.targets.JVMVisitor.emitClosure(JVMVisitor.java:600)
      at org.jruby.ir.targets.JVMVisitor.emitClosures(JVMVisitor.java:592)
      at org.jruby.ir.targets.JVMVisitor.emitScope(JVMVisitor.java:212)
      at org.jruby.ir.targets.JVMVisitor.emitWithSignatures(JVMVisitor.java:554)
      at org.jruby.ir.targets.JVMVisitor.emitMethodJIT(JVMVisitor.java:518)
      at org.jruby.ir.targets.JVMVisitor.codegenScope(JVMVisitor.java:186)
      at org.jruby.ir.targets.JVMVisitor.compileToBytecode(JVMVisitor.java:139)
      at org.jruby.compiler.MethodJITClassGenerator.compile(MethodJITClassGenerator.java:59)
      at org.jruby.compiler.MethodJITTask.exec(MethodJITTask.java:76)
      at org.jruby.compiler.JITCompiler$Task.run(JITCompiler.java:301)
      at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
      at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
      at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
      at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
      at java.base/java.lang.Thread.run(Thread.java:829)
    Caused by: java.lang.IllegalAccessException: no such field: org.jruby.runtime.ThreadContext.runtime/org.jruby.Ruby/getField
      at java.base/java.lang.invoke.MemberName.makeAccessException(MemberName.java:959)
      at java.base/java.lang.invoke.MemberName$Factory.resolveOrFail(MemberName.java:1101)
      at java.base/java.lang.invoke.MethodHandles$Lookup.resolveOrFail(MethodHandles.java:2021)
      at java.base/java.lang.invoke.MethodHandles$Lookup.findGetter(MethodHandles.java:1418)
      at com.headius.invokebinder.Binder.getField(Binder.java:1714)
      at com.headius.invokebinder.Binder.getFieldQuiet(Binder.java:1755)
      ... 27 more
    Caused by: java.lang.LinkageError: loader constraint violation: when resolving field "runtime" of type org.jruby.Ruby, the class loader 'bootstrap' of the current class, java.lang.Object, and the class loader org.apache.catalina.loader.ParallelWebappClassLoader @31f79d66 for the field's defining type, org.jruby.runtime.ThreadContext, have different Class objects for type org.jruby.Ruby (java.lang.Object is in module java.base of loader 'bootstrap'; org.jruby.runtime.ThreadContext is in unnamed module of loader org.apache.catalina.loader.ParallelWebappClassLoader @31f79d66, parent loader java.net.URLClassLoader @2b71fc7e)
      at java.base/java.lang.invoke.MethodHandleNatives.resolve(Native Method)
      at java.base/java.lang.invoke.MemberName$Factory.resolve(MemberName.java:1070)
      at java.base/java.lang.invoke.MemberName$Factory.resolveOrFail(MemberName.java:1098)
      ... 31 more
      
      
  Failing row seems to be https://github.com/jruby/jruby/blob/f7856e414134aaeb0c625c85255fd0736bb9e67c/core/src/main/java/org/jruby/ir/targets/indy/Bootstrap.java#L622

**Comment by headius** [comment]
I think this is probably the same issue I found recently working on JRuby/Maven integration:

https://bugs.openjdk.org/browse/JDK-8313913

This particular case only seemed to affect Java 8; not 11 or higher. Are you running on 8?

**Comment by jlahtinen** [comment]
All applications start without issues if `JAVA_OPTS="-Djruby.jit.max=0"` is added for tomcat command line argument.

**Comment by headius** [comment]
Yeah so it sounds very similar to the OpenJDK issue. I'll see if I can poke some JVM friends to help me investigate further.

**Comment by headius** [comment]
If you are able to reduce this to something I can run locally, it would be an enormous help!

Are you saying you were able to reproduce this issue on Java 11? That might mean it's different than my case, or you've found a more general problem that remains in Java 11.

**Comment by jlahtinen** [comment]
I can try what happens for 2 basic rails application. I'll keep you updated.

And yes. I am using Java 11.

**Comment by jlahtinen** [comment]
[jlahtinen/rails_no_db_jruby_g](https://github.com/jlahtinen/rails_no_db_jruby_g)

There is a project you can use.

Run `script/warble` it generates .war file. Create empty tomcat installation (https://archive.apache.org/dist/tomcat/tomcat-9/v9.0.75/bin/). Remove another webapps (optional). Copy created .war file to webapps e.g. a.war and b.war. start tomcat. After start there should be localhost log file containing stack trace about java.lang.NoClassDefFoundError: Could not initialize class org.jruby.ir.targets.indy.Bootstrap

**Comment by headius** [comment]
@jlahtinen Looking at it now!

**Comment by headius** [comment]
Perhaps I'm doing something wrong? The only errors I see is an initialization error in Rack due to what looks like a hard call to `exit`:

```
11-Sep-2023 10:35:47.735 SEVERE [main] org.apache.catalina.core.ApplicationContext.log ERROR: initialization failed
	org.jruby.rack.RackInitializationException: exit
	from org/jruby/RubyKernel.java:808:in `exit'
	from org/jruby/RubyKernel.java:767:in `exit'
	from uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/bundler/setup.rb:17:in `<main>'
	from org/jruby/RubyKernel.java:1071:in `require'
	from uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/rubygems/core_ext/kernel_require.rb:85:in `require'
	from /Users/headius/work/jruby/tmp/apache-tomcat-9.0.75/webapps/a/WEB-INF/config/boot.rb:3:in `<main>'
	from org/jruby/RubyKernel.java:1071:in `require'
	from uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/rubygems/core_ext/kernel_require.rb:85:in `require'
	from uri:classloader:/jruby/rack/rails/environment3.rb:23:in `load_environment'
	from uri:classloader:/jruby/rack/rails_booter.rb:83:in `load_environment'
	from <script>:1:in `<main>'

		at org.jruby.rack.RackInitializationException.wrap(RackInitializationException.java:29)
		at org.jruby.rack.RackApplicationFactoryDecorator.init(RackApplicationFactoryDecorator.java:104)
		at org.jruby.rack.RackServletContextListener.contextInitialized(RackServletContextListener.java:50)
		at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4494)
		at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:4946)
...
	Caused by: org.jruby.exceptions.SystemExit: (SystemExit) exit
		at org.jruby.RubyKernel.exit(org/jruby/RubyKernel.java:808)
		at org.jruby.RubyKernel.exit(org/jruby/RubyKernel.java:767)
		at RUBY.<main>(uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/bundler/setup.rb:17)
		at org.jruby.RubyKernel.require(org/jruby/RubyKernel.java:1071)
		at RUBY.require(uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/rubygems/core_ext/kernel_require.rb:85)
		at RUBY.<main>(/Users/headius/work/jruby/tmp/apache-tomcat-9.0.75/webapps/a/WEB-INF/config/boot.rb:3)
		at org.jruby.RubyKernel.require(org/jruby/RubyKernel.java:1071)
		at RUBY.require(uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/rubygems/core_ext/kernel_require.rb:85)
		at RUBY.load_environment(uri:classloader:/jruby/rack/rails/environment3.rb:23)
		at RUBY.load_environment(uri:classloader:/jruby/rack/rails_booter.rb:83)
		at RUBY.<main>(<script>:1)
```

Both copies of the app produce the same error. This is running on Tomcat 9.0.75 on Java 1.8ish

**Comment by jlahtinen** [comment]
Try to start tomcat with shell that does not contain rvm ruby installation. e.g. rvm use system before ./bin/startup.sh

**Comment by jlahtinen** [comment]
I was able to reproduce this with Java 1.8ish

11-Sep-2023 14:32:48.841 SEVERE [main] org.apache.catalina.core.StandardContext.listenerStart Exception sending context initialized event to listener instance of class [org.jruby.rack.rails.RailsServletContextListener]
        java.lang.NoClassDefFoundError: Could not initialize class org.jruby.ir.targets.indy.Bootstrap

**Comment by jlahtinen** [comment]
@headius anything new here?

**Comment by headius** [comment]
I will investigate this tomorrow and try to get some fix or workaround in for the upcoming 9.4.4.0 release.

**Comment by headius** [comment]
@jlahtinen I was able to reproduce with your example! Wheels are in motion.

**Comment by headius** [comment]
I see nothing we are doing wrong, so I have added your test case to my JBS report here: https://bugs.openjdk.org/browse/JDK-8313913

It appears to work correctly in Java 17, so it was fixed sometime between 11 and 17. Hopefully we can figure out what was fixed and get it backported, or come up with a workaround.

**Comment by headius** [comment]
If you have any JDK support channels, this might be a good place to use them. Otherwise I think the best workaround I can recommend right now would be to upgrade to something like Java 17.

I will try to see if there's any way to change the way we acquire these method handles that works on 8 and 11.

**Comment by headius** [comment]
@jlahtinen I have pushed https://github.com/jruby/jruby/pull/7949 which appears to fix your simplest case for me locally. This still appears to be a JDK bug, but using a more specific MethodHandles.Lookup object here appears to work correctly.

**Comment by headius** [comment]
@jlahtinen If you could test a build based on https://github.com/jruby/jruby/pull/7949 that would be a big help! `mvn -Pcomplete` will build and install a complete jar, and our nightly builds should push similar artifacts to the Sonatype maven snapshot repository.

**Comment by jlahtinen** [comment]
With jars generated from #7949 the problems seems to be gone.

**Comment by jlahtinen** [comment]
I also found why the example app didn't work for you. You are missing the master key.

To generate a new one, before script/warble run:
1) `rm config/credentials.yml.enc`
2) `EDITOR=vi bin/rails credentials:edit #no need to change anything`

then `script/warble`

I also noticed that generated init.rb does not work. This can be fixed by adding line `Gem.paths = ENV` to META-INF/init.rb file before last line (should be `require 'bundler/shared_helpers'`). I do this by allowing tomcat to extract war contents then editing extracted files and then restarting again (in our software we have custom warbler script phase that adds this line).

**Comment by jlahtinen** [comment]
I you try with jars generated in #7949 you can replaced old ones:
1) ./WEB-INF/lib/jruby-core-9.4.3.0-complete.jar
2) ./WEB-INF/lib/jruby-stdlib-9.4.3.0.jar
with generated ones.

It seems that provided (inside these jruby jars) ffi gem changes version so you have to also change ./WEB-INF/Gemfile.lock to match ffi 1.16.2-java and not ffi 1.15.5-java previously used.

**Comment by jlahtinen** [comment]
@headius with these changes (how to deploy example app) I was able to run 5 rails apps simultaneously in tomcat using jars generated in #7949 without errors.

For me this looks resolved now (JDK bug is a different story). Thank you a lot.

**Comment by headius** [comment]
@jlahtinen That is great news! Hopefully the issue was specific to the field access and we won't see it elsewhere. I will add some comments to my PR commit and get that merged for 9.4.4.0.
